draft-ietf-mls-protocol.md

```diff
@@ -1290,71 +1290,159 @@ Given these inputs, the derivation of secrets for an epoch
+                   init_secret_[n]
 ~~~~~
 
 A number of secrets are derived from the epoch secret for different purposes:
 
-| Secret                | Label         |
-|:----------------------|:--------------|
-| `sender_data_secret`  | "sender data" |
-| `handshake_secret`    | "handshake"   |
-| `application_secret`  | "app"         |
-| `exporter_secret`     | "exporter"    |
-| `confirmation_key`    | "confirm"     |
+| Secret                  | Label         |
+|:------------------------|:--------------|
+| `sender_data_secret`    | "sender data" |
+| `handshake_secret`      | "handshake"   |
+| `application_secret`    | "app"         |
+| `exporter_secret`       | "exporter"    |
+| `confirmation_key`      | "confirm"     |
+| `authentication_secret` | "stateauth"   |
+| `recovery_secret`       | "recovery"    |
 
 ## Pre-Shared Keys
 
 Groups which already have an out-of-band mechanism to generate
-shared group secrets can inject those in the MLS key schedule to seed
+shared group secrets can inject those into the MLS key schedule to seed
 the MLS group secrets computations by this external entropy.
 
-At any epoch, including the initial state, an application can decide
-to synchronize the injection of a PSK into the MLS key schedule.
-
-This mechanism can be used to improve security in the cases where
-having a full run of updates across members is too expensive or in
-the case where the external group key establishment mechanism provides
+The injecting of an external PSK can improve security in the cases
+where having a full run of updates across members is too expensive or in the
+case where the external group key establishment mechanism provides
 stronger security against classical or quantum adversaries.
 
-The security level associated with the PSK injected in the key schedule
-SHOULD match at least the security level of the ciphersuite in use in
-the group.
-
 Note that, as a PSK may have a different lifetime than an update, it
 does not necessarily provide the same Forward Secrecy (FS) or Post-Compromise
 Security (PCS) guarantees as a Commit message.
 
+<!-- OPEN ISSUEClarify lifetime vs security level mandated above. E.g. if
+the PSK security expires before the next update (shorter PSK lifetime than
+update), does that constitute a weaker security level -->
+
+<!-- OPEN ISSUE Define "security level", and what it means to match the
+security level of the ciphersuite used in the group. -->
+
 <!-- OPEN ISSUE: We have to decide if we want an external coordination
 via the application of a Handshake proposal. -->
 
+Each PSK in MLS has a type that designates how it was provisioned.
+External PSKs are provided by the application logic. Other resumption PSKs
+are derived from the MLS key schedule, and used in cases where it is
+necessary to authenticate a member's participation in a prior group state.
+In particular, in addition to external PSK types, a PSK derived from within MLS
+may be used in the following cases:
```

Comment:
Good point in that the types relate their use rather than their provisioning. What do you mean by "at this level"? We do agree that we need strong key separation between the different PSK types, right? That was one of the lessons of various TLS Key Schedule analysis. It's not quite clear to me what you are suggesting we change.

Reply:
I think we do to properly separate the keys depending on their use. This is comparable to TLS, where improper separation between application- and resumption PSKs led to problems down the road. However, we sill only need to derive a single value off of the Key Schedule, as the type is used later when the PSK gets injected into the Key Schedule.